File: moonpoison/back_tracking/17.cpp
#include<bits/stdc++.h>
using namespace std;
class Solution {
public:
    map<char, vector<char>> mp = {
        {'2', {'a', 'b', 'c'}},
        {'3', {'d', 'e', 'f'}},
        {'4', {'g', 'h', 'i'}},
        {'5', {'j', 'k', 'l'}},
        {'6', {'m', 'n', 'o'}},
        {'7', {'p', 'q', 'r', 's'}},
        {'8', {'t', 'u', 'v'}},
        {'9', {'w', 'x', 'y', 'z'}}};
    void backtracking(int i, string digits, string s, vector<string> &ans)
    {
        if(i==digits.length())
        {
            if(s.length()) ans.push_back(s);
            return;
        }
        for(auto C:mp[digits[i]])
            backtracking(i+1, digits, s+C, ans);
    }
    vector<string> letterCombinations(string digits) 
    {
        vector<string> ans;
        backtracking(0, digits, "", ans);
        return ans;
    }
};

File: ice0208/backtracking/17.py
class Solution:
    def __init__(self):
        self.mapping = dict()

        cur = ord("a")
        for i in range(2, 10):
            key = str(i)
            self.mapping[key] = []

            iterate_cnt = 3
            if i in (7, 9):
                iterate_cnt = 4
            for _ in range(iterate_cnt):
                self.mapping[key].append(chr(cur))
                cur += 1

    def __backtracking(self, i: int, digits: str, cur: str):
        if i == len(digits):
            self.answer.append(cur)
            return

        for value in self.mapping[digits[i]]:
            self.__backtracking(i + 1, digits, cur + value)

    def letterCombinations(self, digits: str) -> list[str]:
        if not digits:
            return []

        self.answer = []

        self.__backtracking(0, digits, "")

        return self.answer


if __name__ == "__main__":
    print(Solution().letterCombinations("23"))
    print(Solution().letterCombinations(""))
    print(Solution().letterCombinations("2"))
